### DISCUSSION TITLE
get list of child nodes

### DISCUSSION THREAD

**Discussion by J-Cake** [vizia/vizia#390] [Questions]
Hey, sorry for the rapid-fire questions, but I'm trying to implement a Table widget, as I don't see a straight-forward way of doing this. I'm using a ScrollView, and a custom HStack widget to achieve the table effect. It's custom, so I can pass a Vec of widths to the columns, but now I need to set the widths of each HStack child to the widths defined in the vec, but I don't see a clean way of doing this. 

My table works by constructing it, passing an array of strings as the headers, and a closure which is called repeatedly, until it returns false. Each call is a row. The closure is passed the context, and constructs children within it, like it's the HStack widget.

```rust

struct Row (Rc<Vec<Unit>>); // impls are copied from HStack

ScrollView::new(cx, 0.0, 0.0, false, true, move |cx| {
  Row::new(cx, |cx| {
    (body)(cx);
  });
});
```

I like this approach because it seems to conform with the Vizia examples, and generally seems familiar. However, as I have no control over the closure, I cannot set the widths of each column. Is there a way I can achieve this?

Thanks

**Comment by geom3trik** [comment]
You could do this with binding. The way I would approach this is to have a custom view with a vec for the column widths. The length of this vec would be synced to the number of columns in your data. Then when you construct your rows you could also construct a `HStack` or whatever for each column and then bind the widths of these to the values in the vec in your custom view.

This all hinges on what your data looks like. I'm asuming that it's something like `Vec<Vec<>>`, i.e. a vec or rows and then a vec of columns for each row. I would say this involves some fairly advanced vizia concepts, so I'll try to put together an example for you that demonstrates this better.

**Reply by J-Cake** [reply]
Yep. My data will take roughly that shape, except it'll most probably come from `IntoIterator` structs of some sort, but the concept is the same. 

The only thing I'm not quite clear on is the last sentence of the first paragraph:
> Then when you construct your rows you could also construct a HStack or whatever for each column and then bind the widths of these to the values in the vec in your custom view.

I was hoping to hand over the populating of the columns to the caller, who would construct the table row-wise, rather than cell-wise. This is mostly due to the fact that the cell-wise approach would result in a substantially more complex type signature, and lead to me making errors.

As for your example, I would really appreciate that, as that would give me some sort of anchor from which I can expand my table widget

**Reply by geom3trik** [reply]
You're not wrong that it results in a more complex type signature, but this was the widths can be entirely controlled by the table. I've put together an example of how a table could be done here: <https://github.com/vizia/vizia/blob/table/examples/views/table.rs>. This assumes that the data is `Vec<Vec<T>>`, so a vec of rows and each row has a vec of items with the same number of items for each row.

For my implementation I'm using two `List` views, one for the rows (which is passed in as a lens to the table), and then another for the items. Each item is wrapped in a `HStack` which I'm using to control the column width. The column widths are stored in the table itself and then bound to the `HStacks` with an index lens.

I've tried to keep the example simple so there's a few things it doesn't do which it probably should to be general purpose:
- The widths vector is initialised when the table is made so the number of columns can't be changed after creation of the table. This can be fixed with a binding which sends an event which the table uses to update the number of columns.
- The widths are fixed and the columns can't be resized. This will require some extra view pieces and event handling but it's possible.
- There's no scrollview for when there's too many rows to fit.
- The lists could be virtualised.

The example ended up being quite complex so I'm considering doing a writeup that would take the reader through the steps to build such a table. Could be a good tutorial for some more advanced concepts. Feel free to ask me anything about it though, and in case you're not aware, we have a [Discord ](https://discord.gg/aNkTPsRm2w).
